### ISSUE TITLE
RFC: Add a `move` method, or tweak `insert`

### ISSUE THREAD

**Issue by Victorystick** [Rich-Harris/magic-string#14]
Now that there exists a [Rollup issue for SystemJS](https://github.com/rollup/rollup/issues/83), we'll need to be able to hoist function declarations, while retaining SourceMaps.

I haven't been able to find any `move` method, and the method combination that could have replicated the behavior (`snip`, `remove` and `insert`) requires `insert` to work on MagicStrings in addition to strings.

I suggest extending MagicStrings API with either a `move` method,  or an `insert` that supports MagicStrings.

**Comment by Victorystick** [comment]
The fundamental operation I'm looking for is a way to turn this:
```js
function add( x, y ) {
  return x + y;
}

function sub( x, y ) {
  return x - y;
}
```
into this:
```js
function sub( x, y ) {
  return x - y;
}

function add( x, y ) {
  return x + y;
}
```
while preserving source maps.

**Comment by Rich-Harris** [comment]
This seems doable, though it'd require a bit of a rethink about how mappings are stored. Hmm... I don't quite understand the use case – why would you need to move functions around?

**Comment by Victorystick** [comment]
Ok, the example doesn't really make it clear.

If you take a look at [SystemJS](https://github.com/ModuleLoader/es6-module-loader/wiki/System.register-Explained), what I propose should make more sense. In SystemJS, declarations are hoisted to the top of the wrapper, while all other statements are kept at the "bottom", inside another function.

For example:
```js
import { a } from 'foo';

export var x = a( data, 17 );

export function data(...) { ... }
```
becomes something like:
```js
System.register(['foo'], function($__export) {
    var a, x;
    function data(...) { ... }
    $__export('data', data);
    return {
      setters: [
      // every time a dependency updates an export, 
      // this function is called to update the local binding
      // the setter array matches up with the dependency array above
      function(foo) {
        a = foo.a;
      }
      ],
      execute: function() {
        // use the export function to update the exports of this module
        x = a( data, 17 );
        $__export( 'x', x );
      }
    };
  });
```

So you see, I don't want to change the order of function declarations by moving them around, no. Depending on what you consider being moved where, I want to move Statements down below all declarations.

**Comment by Rich-Harris** [comment]
This is implemented in 0.11! Bit experimental – no sourcemap tests yet, for example – but it seems to work. [Tests have examples of usage](https://github.com/Rich-Harris/magic-string/blob/835b61cae9fc07b6bded38f2f42b38891364e45a/test/index.js#L417-L526) – basically goes like this:

```js
var s = new MagicString( 'abcdefghijkl' );
s.move( 0, 3, 6 ); // start, end, newIndex

assert.equal( s.toString(), 'defabcghijkl' );
```
